homeassistant/config.py

```diff
@@ -1015,21 +1029,115 @@ async def merge_packages_config(
     return config
 
 
+@callback
+def async_process_component_config_errors(
+    hass: HomeAssistant,
+    integration: Integration,
+    config_exception_info: list[ConfigExceptionInfo],
+    raise_on_failure: bool = False,
+) -> None:
+    """Process component configuration errors."""
+
+    p_ex: ConfigExceptionInfo
+    config_error_messages: list[
+        tuple[str, ConfigExceptionInfo, str, str, int | str]
+    ] = []
+    general_error_messages: list[tuple[str, ConfigExceptionInfo]] = []
+    domain = integration.domain
+    for p_ex in config_exception_info:
+        link = p_ex.integration_link or integration.documentation
+        ex = p_ex.ex
+        p_name = p_ex.platform_name
+        p_config = p_ex.config
+        config_file: str = "?"
+        line: int | str = "?"
+        if (log_message := p_ex.log_message) is None:
+            # No pre defined log_message is set, so we generate a message
+            if isinstance(ex, vol.Invalid):
+                log_message = format_schema_error(hass, ex, p_name, p_config, link)
+                if annotation := find_annotation(p_config, ex.path):
+                    config_file, line = annotation
+            elif isinstance(ex, HomeAssistantError):
+                log_message = format_homeassistant_error(
+                    hass, ex, p_name, p_config, link
+                )
+            else:
+                log_message = str(ex)
+            config_error_messages.append((domain, p_ex, log_message, config_file, line))
+        else:
+            general_error_messages.append((domain, p_ex))
+        _LOGGER.error(log_message, exc_info=p_ex.ex if p_ex.log_exception else None)
+
+    if not raise_on_failure or not config_exception_info:
+        return
```

Comment:
This is weird. I'd split this check and check for the list empty immediately at the top of the body. Just check here if we want to raise.